lib/auth/auth.go

```diff
@@ -1322,13 +1322,13 @@ func (a *Server) generateHostCert(
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	if p.Role == types.RoleNode {
-		if lockErr := a.checkLockInForce(authPref.GetLockingMode(),
-			[]types.LockTarget{{Node: p.HostID}, {Node: HostFQDN(p.HostID, p.ClusterName)}},
-		); lockErr != nil {
-			return nil, trace.Wrap(lockErr)
-		}
+
+	if lockErr := a.checkLockInForce(authPref.GetLockingMode(),
```

Comment:
In Teleport, the term `node` almost always refers to an instance of the `ssh_service`.

I am not in favor of making `--node=UUID` work for things that are not SSH nodes.

Can we please add dedicated flags for different types of agents instead?

Reply:
I have mixed feelings about it as well but adding a field per type makes me think that maybe it's too much and if someone forgets to add a type we cannot enforce locks.

What do you think about adding an extra flag `--host-id` and `HostID` into `LockTarget`? 
We can deprecate `--node` in favor of `--host-id` but still support it.

```go

type LockTarget struct {
	// User specifies the name of a Teleport user.
	User string `protobuf:"bytes,1,opt,name=User,proto3" json:"user,omitempty"`
	// Role specifies the name of an RBAC role known to the root cluster.
	// In remote clusters, this constraint is evaluated before translating to local roles.
	Role string `protobuf:"bytes,2,opt,name=Role,proto3" json:"role,omitempty"`
	// Login specifies the name of a local UNIX user.
	Login string `protobuf:"bytes,3,opt,name=Login,proto3" json:"login,omitempty"`
	// Node specifies the UUID of a Teleport node.
	// A matching node is also prevented from heartbeating to the auth server.
	Node string `protobuf:"bytes,4,opt,name=Node,proto3" json:"node,omitempty"`
	// MFADevice specifies the UUID of a user MFA device.
	MFADevice string `protobuf:"bytes,5,opt,name=MFADevice,proto3" json:"mfa_device,omitempty"`
	// WindowsDesktop specifies the name of a Windows desktop.
	WindowsDesktop string `protobuf:"bytes,6,opt,name=WindowsDesktop,proto3" json:"windows_desktop,omitempty"`
	// AccessRequest specifies the UUID of an access request.
	AccessRequest string `protobuf:"bytes,7,opt,name=AccessRequest,proto3" json:"access_request,omitempty"`
	// Device is the device ID of a trusted device.
	// Requires Teleport Enterprise.
	Device               string   `protobuf:"bytes,8,opt,name=Device,proto3" json:"device,omitempty"`
	XXX_NoUnkeyedLiteral struct{} `json:"-"`
	XXX_unrecognized     []byte   `json:"-"`
	XXX_sizecache        int32    `json:"-"`
}

```

Reply:
What would that look like in the UI though?

Reply:
@zmb3 

My idea is to add a HostID field to LockTargets
```diff
// LockTarget lists the attributes of interactions to be disabled.
message LockTarget {
  option (gogoproto.goproto_stringer) = false;
  option (gogoproto.stringer) = false;

  // User specifies the name of a Teleport user.
  string User = 1 [(gogoproto.jsontag) = "user,omitempty"];

  // Role specifies the name of an RBAC role known to the root cluster.
  // In remote clusters, this constraint is evaluated before translating to local roles.
  string Role = 2 [(gogoproto.jsontag) = "role,omitempty"];

  // Login specifies the name of a local UNIX user.
  string Login = 3 [(gogoproto.jsontag) = "login,omitempty"];

  // Node specifies the UUID of a Teleport node.
  // A matching node is also prevented from heartbeating to the auth server.
+  // DEPRECATED: use HostID instead.
  string Node = 4 [(gogoproto.jsontag) = "node,omitempty"];

  // MFADevice specifies the UUID of a user MFA device.
  string MFADevice = 5 [(gogoproto.jsontag) = "mfa_device,omitempty"];

  // WindowsDesktop specifies the name of a Windows desktop.
  string WindowsDesktop = 6 [(gogoproto.jsontag) = "windows_desktop,omitempty"];

  // AccessRequest specifies the UUID of an access request.
  string AccessRequest = 7 [(gogoproto.jsontag) = "access_request,omitempty"];

  // Device is the device ID of a trusted device.
  // Requires Teleport Enterprise.
  string Device = 8 [(gogoproto.jsontag) = "device,omitempty"];

+  // HostID is the host id of the Teleport instance.
+  string HostID = 9 [(gogoproto.jsontag) = "host_id,omitempty"];
}
```

UI and tsh will send locks for nodes, DB, Kube, Apps agents as:
```go
types.LockTarget{HostID:"<uuid_of_the_node>"}
```

We would still support `Node` to keep compatibility with older locks but `HostID` should be the default mode.
```go
types.LockTarget{Node:"<uuid_of_the_node>"}
```

UI already sends the locks with a JSON representation of `types.LockTarget` so the change is simple.
Besides, locking by HostID seems better to lock agents that might have multiple roles (APP, DB,Kube) at the same time and HostID would apply to all of them without having to create multiple locks per type. 

A big difference in the UI is that for kube, app, and db agents the aggregation must happen per host_id. For example, Listing Kubernetes agents:

| Host ID   | Kubernetes Clusters                 |
|-----------|-------------------------------------|
| <host_id_1> | <kube_cluster1>, <kube_cluster2>... |
| <host_id_2> | <kube_cluster1>, <kube_cluster2>... |

We can also include Apps and Dbs that a certain host_id proxies and prompt the user saying: Are you sure you want to block app1, kube1, db1 access for host <host_iD>-


What do you think about it?

PS: need to improve my UX skills

Reply:
Are we expecting any user to be relying on locks on "node" to only apply to the SSH service from a teleport instance but not other services in the same instance? If not, recent `tctl`s should just disallow `--node` and suggest that users specify `--hostid` instead. Otherwise, we could warn about the deprecation of `--node` and recommend `--hostid` instead.

Adding a field to `LockTarget` has the issue that older components in the cluster will see the target as being empty if only `HostID` is set, which would have the side effect of not even locking the SSH services under lock; we can get around that by filling both `Node` and `HostID` for a couple major versions (locks are synced to leaf clusters, which can be one version behind), and considering a nonempty `HostID` to be completely replacing the value in `Node` (to define the ambiguity if the two fields have different values).

Reply:
My idea is to hide the `--node` flag and turn everything into `--host-id` or `--server-id` (I think @fspmarshall used ServerID for the inventory stuff, so I should keep everything consistent and use `server-id`).

Most recent versions of `tctl`
- `tctl lock --node`: fills both `Node` entries. `--node` becomes a hidden flag and tctl lock help would show just `--server-id`. The flag exists to keep UX compatibility but users cannot see it.

- `tctl lock --server-id`: fills both `Node` and `ServerID`  entries. This has no harm since node only applies to Node objects on older Teleport versions. If the agent isn't a node, no lock is applied. For Teleport 15, we stop filling the node and just fill in `ServerID`. Filling both fields aims to keep compatibility with older Teleport versions' lock behavior of SSH servers.

> Adding a field to LockTarget has the issue that older components in the cluster will see the target as being empty if only HostID is set, which would have the side effect of not even locking the SSH services under lock; we can get around that by filling both Node and HostID for a couple of major versions (locks are synced to leaf clusters, which can be one version behind), and considering a nonempty HostID to be completely replacing the value in Node (to define the ambiguity if the two fields have different values).

I am afraid we need to support the deprecated `Node` field forever because customers can have `Node` entries in LockTargets for objects that never expire, which is a problem for us because we cannot convert them to `ServerID` if the agent runs multiple identities.  This is not a big deal since the amount of code isn't considerable.

Reply:
I'm wondering if it's not just something we should warn about in the release notes; I can't think of a reason why you'd want an otherwise working teleport instance locked out on just one service but not the other ones.